shell/platform/linux/fl_json_message_codec.cc

```diff
@@ -0,0 +1,300 @@
+    case FL_VALUE_TYPE_BOOL:
+      writer.Bool(fl_value_get_bool(value));
+      break;
+    case FL_VALUE_TYPE_INT:
+      writer.Int64(fl_value_get_int(value));
+      break;
+    case FL_VALUE_TYPE_FLOAT:
+      writer.Double(fl_value_get_float(value));
+      break;
+    case FL_VALUE_TYPE_STRING:
+      writer.String(fl_value_get_string(value));
+      break;
+    case FL_VALUE_TYPE_UINT8_LIST: {
+      writer.StartArray();
+      const uint8_t* data = fl_value_get_uint8_list(value);
+      for (size_t i = 0; i < fl_value_get_length(value); i++)
+        writer.Int(data[i]);
+      writer.EndArray();
+      break;
+    }
+    case FL_VALUE_TYPE_INT32_LIST: {
+      writer.StartArray();
+      const int32_t* data = fl_value_get_int32_list(value);
+      for (size_t i = 0; i < fl_value_get_length(value); i++)
+        writer.Int(data[i]);
+      writer.EndArray();
+      break;
+    }
+    case FL_VALUE_TYPE_INT64_LIST: {
+      writer.StartArray();
+      const int64_t* data = fl_value_get_int64_list(value);
+      for (size_t i = 0; i < fl_value_get_length(value); i++)
+        writer.Int64(data[i]);
+      writer.EndArray();
+      break;
+    }
+    case FL_VALUE_TYPE_FLOAT_LIST: {
+      writer.StartArray();
+      const double* data = fl_value_get_float_list(value);
+      for (size_t i = 0; i < fl_value_get_length(value); i++)
+        writer.Double(data[i]);
+      writer.EndArray();
+      break;
+    }
+    case FL_VALUE_TYPE_LIST:
+      writer.StartArray();
+      for (size_t i = 0; i < fl_value_get_length(value); i++)
+        if (!write_value(writer, fl_value_get_list_value(value, i), error))
+          return FALSE;
+      writer.EndArray();
+      break;
+    case FL_VALUE_TYPE_MAP:
+      writer.StartObject();
+      for (size_t i = 0; i < fl_value_get_length(value); i++) {
+        FlValue* key = fl_value_get_map_key(value, i);
+        if (fl_value_get_type(key) != FL_VALUE_TYPE_STRING) {
+          g_set_error(error, FL_JSON_MESSAGE_CODEC_ERROR,
+                      FL_JSON_MESSAGE_CODEC_ERROR_INVALID_OBJECT_KEY_TYPE,
+                      "Invalid object key type");
+          return FALSE;
+        }
+        writer.Key(fl_value_get_string(key));
+        if (!write_value(writer, fl_value_get_map_value(value, i), error))
+          return FALSE;
+      }
+      writer.EndObject();
+      break;
+    default:
+      g_set_error(error, FL_MESSAGE_CODEC_ERROR, FL_MESSAGE_CODEC_ERROR_FAILED,
+                  "Encountered unknown FlValue type");
+      return FALSE;
+  }
+
+  return TRUE;
+}
+
+// Handler to parse JSON using rapidjson in SAX mode
+struct FlValueHandler {
+  GPtrArray* stack;
+  FlValue* key;
+
+  FlValueHandler() {
+    stack = g_ptr_array_new_with_free_func(
+        reinterpret_cast<GDestroyNotify>(fl_value_unref));
+    key = nullptr;
+  }
+
+  ~FlValueHandler() {
+    g_ptr_array_unref(stack);
+    if (key != nullptr)
+      fl_value_unref(key);
+  }
+
+  // Get the current head of the stack
```

Comment:
Gets.

Also, all your comments are missing periods (style requires this, it's not just me :) )

Reply:
One day I'll internalize the grammar...